### DISCUSSION TITLE
Creating and implementing custom native modules based on third-party services.

### DISCUSSION THREAD

**Discussion by swamptoday** [expo/expo#24361] [Archived: Questions]
I am developing application that needs to have the payment functionality. 
I need to use third-party payment service - they offer Android and IOS SDKs. I am now creating custom native module, based on their documentation. However I am a total beginner and I even do not know where I should start.

For example: 
1. I have a page where user is supposed to choose service and press "Pay button", when ready. How do I connect this button with future native code and how to send pay values to native module?
2. The payment functionality is not the only one - creating tokens for future card usage is also a need (third-party service offers it too). How do I implement different usage of module then?
3. The service itself redirects user on their pay page in the application after sending request of payment, so do I need to have a ExpoNameView.kt(swift) files or only ExpoNameModule.kt(swift) files?
4. How to send back results from native module?
5. Can I write Android module using Java (documentation of service is written using Java, however 'create-expo-module' creates files View and Module in Kotlin)?

I will be very grateful if someone can explain these and, maybe, share some detailed documentation or examples of this usage of custom native modules based on third-party services. Because the only documentation I was able to find [Use third-party native libraries in Expo](https://docs.expo.dev/modules/third-party-library/) does not fully cover my questions - only the implementation parts. 

Thanks everyone in advance!

**Comment by eriktoor** [comment]
Hey @swamptoday did you ever figure out how to do this? I am working to create a payment module with a payment processor other than Stripe.

**Reply by swamptoday** [reply]
Hi @eriktoor!

Yes, I actually figured it out, but I implemented it using react native custom native module, as I have not mamaged to find any usefull information on expo native modules.

However, your solution depends on what technology you have to/want to use. If you are free to use any - finding information on react native modules is easier. Also, you can try to find whether there is already a react-native module for your third-party service. I don't know if this would bring any help, but I've actually found my service in react-native modules: https://github.com/kyivstarteam/react-native-ecom-portmone. Maybe this will help you to see the basic structure.

But, if you have to use only expo, then, unfortunately, I will not be able to give any other advice, than to contact specialist in expo custom native modules or finding more tutorials.

Hope you will succeed in creating your module!

**Reply by eriktoor** [reply]
thanks for sharing this information! I have embraced the idea of using modules / plugins / native code to extend what expo already has. just out of curiosity what is portmone?

**Reply by swamptoday** [reply]
It is payment service, that you can use in Web/android sdk/ios sdk. But you have to register as their merchant and pay for its usage.

**Reply by eriktoor** [reply]
Oh that's awesome! Does it do apple pay / google pay? If it does apple pay I might apple and just leverage this!! Looks really cool.

**Reply by swamptoday** [reply]
Yes, it actually let's you implement Google and Apple pay for Web, Android and iOS

**Reply by eriktoor** [reply]
that's awesome! where can i apply to use Portmone?

**Reply by swamptoday** [reply]
Sorry for taking so long to answer!
I am not familiar with the whole detailed process of applying, but i think you can find it all on their website https://www.portmone.com.ua/ecommerce.
It is fully Ukrainian, but I think it would be no problem for you to use translator for websites or just directly contact them in language that is suitable for you. Have a good day!
